plutus-tx-plugin/src/PlutusTx/Compiler/Trace.hs

```diff
@@ -0,0 +1,41 @@
+{-# LANGUAGE FlexibleContexts  #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module PlutusTx.Compiler.Trace where
+
+import PlutusTx.Compiler.Types
+import PlutusTx.Compiler.Utils
+
+import Control.Monad.Extra
+import Control.Monad.Reader
+import Control.Monad.State
+import Data.Maybe
+import Debug.Trace
+import GHC.Plugins qualified as GHC
+
+traceCompilation ::
+  (MonadReader (CompileContext uni fun) m, MonadState CompileState m) =>
+  -- | The thing (expr, type, kind, etc.) being compiled
+  GHC.SDoc ->
+  -- | The compilation action
+  m a ->
+  m a
+traceCompilation sd compile = ifM (notM (asks ccDebugTraceOn)) compile $ do
+  CompileState nextStep prevSteps <- get
+  put $ CompileState (nextStep + 1) (nextStep : prevSteps)
+  let mbParentStep = listToMaybe prevSteps
+  s <- sdToStr sd
+  traceM $
```

Comment:
This stuff starts to make me wonder if we should just be emitting logs in some more standard format that includes this kind of information, like https://hackage.haskell.org/package/opentelemetry-0.8.0 or something.

Reply:
Perhaps for this too, but I'd love to have some observability on the scripts being evaluated on mainnet.